File: PaperPiano/js/note-recognizer.js
/**
 * NoteRecognizer - OCR-based musical-note recognition using Tesseract.js.
 * Falls back to sequential default notes when OCR is unavailable or uncertain.
 */

const VALID_NOTES = [
    'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B',
    'Db', 'Eb', 'Gb', 'Ab', 'Bb'
];

const DEFAULT_SCALE = ['C', 'D', 'E', 'F', 'G', 'A', 'B'];

export class NoteRecognizer {
    constructor() {
        this.worker = null;
        this.ready = false;
    }

    async init(onProgress) {
        try {
            if (onProgress) onProgress('Loading OCR engine…');

            // Tesseract is loaded globally via <script>
            if (typeof Tesseract === 'undefined') {
                console.warn('Tesseract.js not loaded; OCR unavailable');
                return;
            }

            // Wrap Tesseract init with a timeout to prevent blocking on slow mobile connections
            const timeoutMs = 5000;
            const initPromise = (async () => {
                this.worker = await Tesseract.createWorker('eng', 1, {
                    logger: m => {
                        if (m.status === 'loading tesseract core' || m.status === 'initializing tesseract') {
                            if (onProgress) onProgress(`OCR: ${m.status}`);
                        }
                    }
                });

                await this.worker.setParameters({
                    tessedit_char_whitelist: 'ABCDEFGabcdefg#',
                    tessedit_pageseg_mode: '8'   // PSM.SINGLE_WORD
                });
            })();

            const timeoutPromise = new Promise((_, reject) =>
                setTimeout(() => reject(new Error('OCR init timed out')), timeoutMs)
            );

            await Promise.race([initPromise, timeoutPromise]);

            this.ready = true;
            if (onProgress) onProgress('OCR ready ✓');
        } catch (err) {
            console.warn('NoteRecognizer init failed (OCR optional):', err);
            if (onProgress) onProgress('OCR skipped (slow connection)');
        }
    }

    /**
     * Attempt to recognize a note name from a cropped canvas region.
     * @param {HTMLCanvasElement} canvas - pre-processed crop
     * @returns {string|null} recognized note or null
     */
    async recognize(canvas) {
        if (!this.ready || !this.worker) return null;
        try {
            const { data } = await this.worker.recognize(canvas);
            return this.parse(data.text);
        } catch {
            return null;
        }
    }

    /**
     * Parse raw OCR text into a valid musical note name.
     */
    parse(text) {
        if (!text) return null;

        let cleaned = text.trim().toUpperCase().replace(/[^A-G#]/g, '');
        if (!cleaned) return null;

        const letter = cleaned.charAt(0);
        if (!'ABCDEFG'.includes(letter)) return null;

        let note = letter;
        if (cleaned.length > 1 && cleaned.charAt(1) === '#') {
            note += '#';
        }

        return VALID_NOTES.includes(note) ? note : (VALID_NOTES.includes(letter) ? letter : null);
    }

    /**
     * Return a sensible default note for a given shape index.
     */
    getDefaultNote(index) {
        return DEFAULT_SCALE[index % DEFAULT_SCALE.length];
    }

    destroy() {
        this.worker?.terminate();
    }
}

File: PaperPiano/note_recognizer.py
"""
NoteRecognizer — OCR-based musical-note recognition using pytesseract.
Falls back to sequential default notes when OCR is unavailable or uncertain.
"""

import re

VALID_NOTES = [
    'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B',
    'Db', 'Eb', 'Gb', 'Ab', 'Bb',
]

DEFAULT_SCALE = ['C', 'D', 'E', 'F', 'G', 'A', 'B']

# Try to import pytesseract
_tesseract_available = False
try:
    import pytesseract
    _tesseract_available = True
except ImportError:
    pass


class NoteRecognizer:
    def __init__(self):
        self.ready = _tesseract_available

    def recognize(self, image) -> str | None:
        """
        Attempt to recognize a note name from a preprocessed image (numpy array).
        Returns recognized note or None.
        """
        if not self.ready:
            return None
        try:
            text = pytesseract.image_to_string(
                image,
                config='--psm 8 -c tessedit_char_whitelist=ABCDEFGabcdefg#',
            )
            return self.parse(text)
        except Exception:
            return None

    def parse(self, text: str) -> str | None:
        """Parse raw OCR text into a valid musical note name."""
        if not text:
            return None

        cleaned = re.sub(r'[^A-Ga-g#]', '', text.strip().upper())
        if not cleaned:
            return None

        letter = cleaned[0]
        if letter not in 'ABCDEFG':
            return None

        note = letter
        if len(cleaned) > 1 and cleaned[1] == '#':
            note += '#'

        if note in VALID_NOTES:
            return note
        if letter in VALID_NOTES:
            return letter
        return None

    def get_default_note(self, index: int) -> str:
        """Return a sensible default note for a given shape index."""
        return DEFAULT_SCALE[index % len(DEFAULT_SCALE)]
